### ISSUE TITLE
Make TreeBuilder accessible from outside the dTree package

### ISSUE THREAD

**Issue by microbecode** [ErikGartner/dTree#35]
I wanted to try your framework and followed the installation instructions. I referenced d3, lodash and dTree. Created the options and data and tried to initialize. I get error about missing TreeBuilder.

I guess the TreeBuilder is supposed to be included in the https://cdn.rawgit.com/ErikGartner/dTree/1.3.1/dist/dTree.min.js or then it needs to be separately referenced?

**Comment by ErikGartner** [comment]
Hi @microbecode!

No it should be included as you mention.

- What browser are you running?
- Can you attach a log file?
- Is the website public, can you link to it or share a https://jsfiddle.net/?
- Does the non-minified version work?

**Comment by microbecode** [comment]
Running Chrome. But surely the problem isn't dependent on the browser. If a file is missing it's missing.
https://jsfiddle.net/09aud823/ .
Non-minified version doesn't work either. It seems to have the TreeBuilder but possibly not accessible or so.

I don't have npm installed (yet) so haven't tried the 'real' installation.

**Comment by ErikGartner** [comment]
I have tested your code and there are two problems.

Firstly you include d3 v4 while dTree currently supports v3. This is a separate matter and I've created a ticket for this.

Secondly, yes TreeBuilder isn't readily accessible for outside the dTree package. While this is something that I should fix it isn't really a problem in most situation since it contains code that is used internally.

Notice that the example that you run isn't really a good one since the options object that you use is the internal default state. There is no need to overwrite all the callback with TreeBuilder calls, this is default mode. You should only overwrite callback if they do not intend to use the TreeBuilder functions.

So basically don't use TreeBuilder in the options object and you should be fine. Have a look at the demo example I mentioned before it is a better starting point and works fine.

**Comment by microbecode** [comment]
Thanks for reply.
I had no idea what the TreeBuilder is - I just took the example code and smashed it all together. Maybe the documentation could be improved on that part.
And yes, didn't notice that I used wrong d3 version.

**Comment by ErikGartner** [comment]
I hoped this solved your problem.

I admit that now that I look at it the default options object in the readme looks a lot like a suitable example.  So yeah, improving documentation probably wouldn't hurt!

/Erik

**Comment by daddie888** [comment]
Is there a working sample on jsfiddle please ?

**Comment by ErikGartner** [comment]
No, but look in [test/demo](https://github.com/ErikGartner/dTree/tree/master/test/demo) for a simple example.

**Comment by daddie888** [comment]
Thanks, but when I open the index.html I get a cross domain error on the
json.
Made one myself and put it on jsfiddle
https://jsfiddle.net/daddie888/Lz8cv1z4/
You can use it if you like.
Would you like to take a look at https://jsfiddle.net/daddie888/y934abmc/1/
also ?
I would like to avoid the elbow above Peter. Is that possible ? In the
other example it dispappears when there are more children/spouces, but what
when there are only two ?

Greeting

Peter

2017-01-05 21:40 GMT+01:00 Erik Gärtner <notifications@github.com>:

**Comment by ErikGartner** [comment]
Hi Peter!

Thanks for creating the JSFiddle. I probably should make an official one, the demo isn't as convenient as it should be.

The elbow is an unfortunate consequence of the layout engine used by dTree. Until it is replaced with a more advanced algorithm you'll get the elbow for two nodes like in your example.

dTree uses the [D3 tree layout](https://github.com/d3/d3-3.x-api-reference/blob/master/Tree-Layout.md). I'm pretty busy with other projects right now. Feel free to hack away at a better implementation and in that case feel free to make a pull request!

Best regards,
Erik
